hail/src/main/scala/is/hail/types/encoded/EArray.scala

```diff
@@ -100,22 +100,69 @@ final case class EArray(val elementType: EType, override val required: Boolean =
 
     val len = cb.newLocal[Int]("len", in.readInt())
     val array = cb.newLocal[Long]("array", arrayType.allocate(region, len))
+    val elemAddr = cb.newLocal[Long]("elemAddr", arrayType.firstElementOffset(array, len))
     arrayType.storeLength(cb, array, len)
 
     val i = cb.newLocal[Int]("i")
+    val mbyteOffset = cb.newLocal[Long]("mbyteOffset", array + arrayType.lengthHeaderBytes)
+    val mbyte = cb.newLocal[Byte]("mbyte", 0.toByte)
+    val presentBitsLong = cb.newLocal[Long]("presentBitsLong", 0)
+    val inBlockIndexToPresentValue = cb.newLocal[Int]("inBlockIndexToPresentValue", 0)
     val readElemF = elementType.buildInplaceDecoder(arrayType.elementType, cb.emb.ecb)
 
     if (!elementType.required)
       cb += in.readBytes(region, array + const(arrayType.lengthHeaderBytes), arrayType.nMissingBytes(len))
 
-    cb.forLoop(cb.assign(i, 0), i < len, cb.assign(i, i + 1), {
-      val elemAddr = cb.memoize(arrayType.elementOffset(array, len, i))
-      if (elementType.required)
-        readElemF(cb, region, elemAddr, in)
-      else
-        cb.ifx(arrayType.isElementDefined(array, i),
-          readElemF(cb, region, elemAddr, in))
-    })
+    val LONG_MASK_ALL_BUT_FIRST = const(0x7fffffffffffffffL)
+
+    def numberOfLeadingZeros(l: Code[Long]): Code[Int] =
+      Code.invokeStatic1[java.lang.Long, Long, Int]("numberOfLeadingZeros", l)
```

Comment:
Btw, we already have `Code[Long].numberOfLeadingZeros` (and trailing too)

Reply:
I am crushed to learn this is completely unused elsewhere.

Reply:
Fixed.

Reply:
I’m using pop count, I think in the rng. Don’t remember if I used leading zeroes at some point, or if I just added them at the same time as pop count out of thoroughness.